rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,380 @@
+```
+* 'someinput': 'git+https://github.com/some/input.git?ref=master&rev=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' -> 'git+https://github.com/some/input.git?ref=master&rev=bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb' (Untrusted newer commits are available)
+```
+In most cases the above situation is likely only temporary, and due to
+the untrusted commits being ignored, if the situation is resolved upstream,
+downstream users will have no problem and won't have to pay attention to it.
+
+## Speeding up Git
+[speeding up git](#speeding-up-git)
+
+Taking the below benchmarks into account, we want to make Nix
+by default fetch with `--filter=tree:0`.
+
+### Rationale
+
+Cryptographically verifying Git commits is easy, but
+Git is currently often not suitable for fetching in Nix.
+Nixpkgs is growing at an extreme rate, and with each
+revision cloning it becomes slower.
+For the author, a full clone from scratch takes over 4 minutes!
+
+Because of this, fetching the tarball through GitHub's APIs
+is preferred, as it is much faster (around 60 times for the author).
+
+Since cryptographic verification of updates of repositories
+through tarballs coming from GitHub is much harder,
+we must fetch the Git commits instead so that we can verify them.
+To speed up fetching Git commits, we can make use of relatively new Git features
+to only download the parts we need:
+- https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone/
+- https://git-scm.com/docs/partial-clone
+- https://git-scm.com/docs/shallow
+- See man page for `git-clone(1)` and `git-rev-list(1)` and search for `--filter`
+
+A benchmark from the author of this RFC:
+- `git clone git://github.com/NixOS/nixpkgs --single-branch --branch=21.05-beta` 4:17
+- `git clone git://github.com/NixOS/nixpkgs --single-branch --branch=21.05-beta --filter=blob:none` 1:05
+- `git clone git://github.com/NixOS/nixpkgs --single-branch --branch=21.05-beta --filter=tree:0` 0:22
+- `git clone git://github.com/NixOS/nixpkgs --single-branch --branch=21.05-beta --depth 1` 0:08
+  (Almost supported already, see https://github.com/NixOS/nix/issues/5119)
+- `curl https://github.com/NixOS/nixpkgs/archive/21.05-beta.tar.gz -LO | tar -xf -` 0:04
+
+Then after doing `cd nixpkgs`, we try fetching `21.05-beta`, which is 659 commits more to
+pull:
+- `git pull --ff-only origin 21.05` 0:08
+- `git pull --ff-only origin 21.05` 0:08 (after having fetched with `--filter=blob:none`)
+- `git pull --ff-only origin 21.05` 0:07 (after having fetched with `--filter=tree:0`)
+- `git clone git://github.com/NixOS/nixpkgs --single-branch --branch=21.05 --depth 1` 0:08 (pulling doesn't work)
+- `curl https://github.com/NixOS/nixpkgs/archive/21.05.tar.gz -LO | tar -xf -` 0:04
+
+This is with an 850 EVO SSD, the average download speed from GitHub was around 11 MiB/s according to `git`.
+
+While `--depth 1` is faster, but due to its nature it's not clear how it interacts with PGP signing.
```

Comment:
https://git-scm.com/book/en/v2/Git-Internals-Git-Objects if I understand this correctly a commit contains the current tree hash and therefore this should also work. But this would really need to be verified.

Reply:
You can still verify the *signature* I think, but the problem is, I don't know if Git can derive the checksum of the shallow repository's contents from that commit.